aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,468 @@
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+                ciphertext.as_mut_ptr(),
+                ciphertext.len(),
+            ) {
+                return Err(error_value);
+            }
+            kdf(&shared_secret)
+        }
+    }
+
+    /// Creates a new KEM private key from raw bytes. This method is NOT meant to generate
+    /// a new private key, rather it restores a `KemPrivateKey` that was previously converted
+    /// to raw bytes.
+    ///
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    ///
+    pub fn new(alg: &'static KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        match bytes.len().cmp(&alg.secret_key_size()) {
+            Ordering::Less => Err(KeyRejected::too_small()),
+            Ordering::Greater => Err(KeyRejected::too_large()),
+            Ordering::Equal => Ok(()),
+        }?;
+        unsafe {
+            let privkey = LcPtr::new(EVP_PKEY_kem_new_raw_secret_key(
+                alg.id.nid(),
+                bytes.as_ptr(),
+                bytes.len(),
+            ))?;
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                pkey: privkey,
+                priv_key: bytes.into(),
+            })
+        }
+    }
+}
+
+impl Drop for KemPrivateKey {
+    fn drop(&mut self) {
+        self.priv_key.zeroize();
+    }
+}
+
+impl AsRef<[u8]> for KemPrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a `KemPrivateKey` or constructed from raw bytes.
+#[derive(Debug)]
+pub struct KemPublicKey {
+    algorithm: &'static KemAlgorithm,
+    pkey: LcPtr<*mut EVP_PKEY>,
+    pub_key: Box<[u8]>,
+}
+
+impl KemPublicKey {
+    /// Return the algorithm associated with the given KEM public key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static KemAlgorithm {
+        self.algorithm
+    }
+
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    ///
+    /// After the encapsulation is finished, `encapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    #[allow(clippy::missing_panics_doc)]
+    pub fn encapsulate<F, R, E>(&self, error_value: E, kdf: F) -> Result<R, E>
+    where
+        F: FnOnce(&[u8], &[u8]) -> Result<R, E>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut()));
+            if ctx.is_err() {
+                return Err(error_value);
+            }
```

Comment:
Use a let-else here as well to avoid `unwrap`:
```
    let Ok(ctx) =
        LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut())) else {
        return Err(error_value);
    };
```